File: src/main/java/core/gui/event/ChangeEventHandler.java
package core.gui.event;

import javax.swing.event.*;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

/**
 * Support for handling {@link ChangeListener}s.
 */
public class ChangeEventHandler {

    /** CopyOnWriteArrayList ensures thread-safety */
    private final List<ChangeListener> changeListeners = new CopyOnWriteArrayList<>();

    public void addChangeListener(ChangeListener changeListener) {
        changeListeners.add(changeListener);
    }

    public void removeChangeListener(ChangeListener changeListener) {
        if (changeListeners.contains(changeListener)) {
            changeListeners.remove(changeListener);
        }
    }

    public void fireChangeEvent(ChangeEvent event) {
        for (ChangeListener listener : changeListeners) {
            listener.stateChanged(event);
        }
    }
}

File: src/main/java/core/gui/event/ChangeEventHandler.kt
package core.gui.event

import java.util.concurrent.CopyOnWriteArrayList
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener

/**
 * Support for handling [ChangeListener]s.
 */
open class ChangeEventHandler {
    /** CopyOnWriteArrayList ensures thread-safety  */
    private val changeListeners: MutableList<ChangeListener> = CopyOnWriteArrayList()
    fun addChangeListener(changeListener: ChangeListener) {
        changeListeners.add(changeListener)
    }

    fun removeChangeListener(changeListener: ChangeListener) {
        if (changeListeners.contains(changeListener)) {
            changeListeners.remove(changeListener)
        }
    }

    protected fun fireChangeEvent(event: ChangeEvent?) {
        for (listener in changeListeners) {
            listener.stateChanged(event)
        }
    }
}
